packages/common/src/DeployedModule.ts

```diff
@@ -0,0 +1,115 @@
+import {
+    AccountAddress,
+    AccountSigner,
+    AccountTransactionType,
+    CcdAmount,
+    ConcordiumGRPCClient,
+    ContractTransactionMetadata,
+    HexString,
+    InitContractPayload,
+    TransactionExpiry,
+    VersionedModuleSource,
+    getContractUpdateDefaultExpiryDate,
+    signTransaction,
+} from '.';
+import { checkParameterLength } from './contractHelpers';
+import { ModuleReference } from './types/moduleReference';
+import { Buffer } from 'buffer/';
+
+/**
+ * An update transaction without header.
+ */
+export type ContractInitTransaction = {
+    /** The type of the transaction, which will always be of type {@link AccountTransactionType.InitContract} */
+    type: AccountTransactionType.InitContract;
+    /** The payload of the transaction, which will always be of type {@link InitContractPayload} */
+    payload: InitContractPayload;
+};
+
+/** Type representing a smart contract module deployed on chain. */
+export class DeployedModule<C extends string = string> {
+    /** Private constructor to enforce creating objects using a static method. */
+    private constructor(
+        /** The gRPC connection used by this object */
+        public grpcClient: ConcordiumGRPCClient,
+        /** The reference for this module */
+        public moduleReference: ModuleReference
+    ) {}
+
+    /** Create a new `GenericModule` instance for interacting with a smart contract module on chain.
+     * This function ensures the module is already deployed on chain otherwise produces an error. */
+    public static async create(
+        grpcClient: ConcordiumGRPCClient,
+        moduleReference: ModuleReference
+    ): Promise<DeployedModule> {
+        const mod = new DeployedModule(grpcClient, moduleReference);
+        await mod.checkOnChain();
+        return mod;
+    }
+
+    /** Create a new `GenericModule` instance for interacting with a smart contract module on chain.
+     * The caller must ensure that the smart contract module is already deployed on chain. */
+    public static createUnchecked(
+        grpcClient: ConcordiumGRPCClient,
+        moduleReference: ModuleReference
+    ): DeployedModule {
+        return new DeployedModule(grpcClient, moduleReference);
+    }
+
+    /** Check if this module is deployed to the chain. */
+    public async checkOnChain(blockHash?: string): Promise<void> {
```

Comment:
Would it make more sense to have the signature be

```suggestion
    public async checkOnChain(blockHash?: string): Promise<boolean> {
```

?

Maybe also add a `@throws` declaration for GRPC errors?

Reply:
The reason for throwing instead of the boolean is to be able to communicate different reasons for failing the check. This is mainly needed in `checkOnChain` for the `GenericContract`, so this is just to follow the signature